File: udp_server.py
#!/usr/bin/env python3
"""
UDP Server with multiple functionalities
Implements various features from the project requirements
"""

import socket
import threading
import time
import random
import json
import os
from datetime import datetime
from concurrent.futures import ThreadPoolExecutor

class UDPServer:
    def __init__(self, host='localhost', port=9876):
        self.host = host
        self.port = port
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.connected_users = {}
        self.database = {
            "1": "Ahmad Karimov - Dasturchi",
            "2": "Malika Toshmatova - Dizayner", 
            "3": "Javlon Umarov - Menejer",
            "4": "Dilnoza Rahimova - Marketing",
            "5": "Bobur Ismoilov - Texnik"
        }
        self.log_file = "log.txt"
        self.running = True
        
    def start(self):
        """Start the UDP server"""
        try:
            self.socket.bind((self.host, self.port))
            print(f"UDP Server ishga tushdi! Port: {self.port}")
            print("Server turli xil xizmatlarni qo'llab-quvvatlaydi:")
            print("- Oddiy xabar almashish")
            print("- Chat dasturi")
            print("- Matematik amallar")
            print("- Fayl mavjudligini tekshirish")
            print("- Tasodifiy son generatsiya")
            print("- Sensor ma'lumotlari")
            print("- Ma'lumotlar bazasi bilan ishlash")
            print("- Va boshqalar...")
            print()
            
            # Start background service
            self.start_background_service()
            
            # Main server loop
            while self.running:
                try:
                    data, addr = self.socket.recvfrom(1024)
                    client_message = data.decode('utf-8').strip()
                    
                    # Log the request
                    self.log_request(client_message, addr)
                    
                    # Process message in a separate thread
                    threading.Thread(
                        target=self.handle_client,
                        args=(client_message, addr),
                        daemon=True
                    ).start()
                    
                except Exception as e:
                    if self.running:
                        print(f"Xatolik: {e}")
                        
        except Exception as e:
            print(f"Server ishga tushirishda xatolik: {e}")
        finally:
            self.socket.close()
    
    def handle_client(self, message, addr):
        """Handle client message and send response"""
        try:
            # Register user
            user_key = f"{addr[0]}:{addr[1]}"
            self.connected_users[user_key] = addr
            
            # Process message and get response
            response = self.process_message(message, addr)
            
            # Send response back to client
            self.socket.sendto(response.encode('utf-8'), addr)
            
            print(f"Kliyent: {addr} - {message}")
            print(f"Javob: {response}")
            print()
            
        except Exception as e:
            print(f"Kliyent bilan ishlashda xatolik: {e}")
    
    def process_message(self, message, addr):
        """Process incoming messages and return appropriate responses"""
        message_lower = message.lower()
        
        if message_lower in ['salom', 'hello']:
            return "Salom! Xush kelibsiz! Server ishlamoqda."
        
        elif message_lower == 'status':
            return f"Server ishlamoqda. Faol foydalanuvchilar: {len(self.connected_users)}"
        
        elif message_lower == 'ip manzilim qanday?':
            return f"Sizning IP manzilingiz: {addr[0]}"
        
        elif message_lower == 'foydalanuvchilar':
            return f"Faol foydalanuvchilar: {len(self.connected_users)} ta"
        
        elif message_lower == 'vaqt':
            return f"Hozirgi vaqt: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}"
        
        elif message_lower == 'tasodifiy son':
            random_num = random.randint(1, 100)
            return f"Tasodifiy son (1-100): {random_num}"
        
        elif message_lower == 'haroratni yuboring':
            temperature = random.randint(10, 50)
            return f"Harorat: {temperature}°C"
        
        elif message_lower == 'menga maqol ayting':
            return "Maqol: 'Bilim - kuchdir. Har bir kuni yangi narsa o'rganing.'"
        
        elif message.startswith('fayl:'):
            filename = message[5:].strip()
            if os.path.exists(filename):
                size = os.path.getsize(filename)
                return f"Fayl '{filename}' mavjud. Hajmi: {size} bayt"
            else:
                return f"Fayl '{filename}' topilmadi"
        
        elif message.startswith('id:'):
            id_key = message[3:].strip()
            if id_key in self.database:
                return f"Ma'lumot (ID: {id_key}): {self.database[id_key]}"
            else:
                return f"ID '{id_key}' bo'yicha ma'lumot topilmadi"
        
        elif message_lower in ['tosh', 'qaychi', 'qog\'oz']:
            choices = ['tosh', 'qaychi', 'qog\'oz']
            server_choice = random.choice(choices)
            return self.play_rock_paper_scissors(message_lower, server_choice)
        
        elif any(op in message for op in ['+', '-', '*', '/']):
            return self.calculate_math_expression(message)
        
        elif message_lower == 'json':
            data = {
                "foydalanuvchi": addr[0],
                "vaqt": datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
                "status": "faol"
            }
            return json.dumps(data, ensure_ascii=False)
        
        else:
            # Default response - echo with timestamp
            return f"Sizning xabaringiz: '{message}' - {datetime.now().strftime('%H:%M:%S')}"
    
    def calculate_math_expression(self, expression):
        """Calculate mathematical expressions"""
        try:
            if '+' in expression:
                parts = expression.split('+')
                if len(parts) == 2:
                    a = int(parts[0].strip())
                    b = int(parts[1].strip())
                    return f"Natija: {a + b}"
            elif '-' in expression:
                parts = expression.split('-')
                if len(parts) == 2:
                    a = int(parts[0].strip())
                    b = int(parts[1].strip())
                    return f"Natija: {a - b}"
            elif '*' in expression:
                parts = expression.split('*')
                if len(parts) == 2:
                    a = int(parts[0].strip())
                    b = int(parts[1].strip())
                    return f"Natija: {a * b}"
            elif '/' in expression:
                parts = expression.split('/')
                if len(parts) == 2:
                    a = int(parts[0].strip())
                    b = int(parts[1].strip())
                    if b != 0:
                        return f"Natija: {a / b}"
                    else:
                        return "Xatolik: Nolga bo'lish mumkin emas"
        except Exception as e:
            return f"Xatolik: Noto'g'ri matematik ifoda - {e}"
        
        return "Xatolik: Noto'g'ri format"
    
    def play_rock_paper_scissors(self, client_choice, server_choice):
        """Play Rock-Paper-Scissors game"""
        if client_choice == server_choice:
            return f"Durrang! Siz: {client_choice}, Server: {server_choice}"
        elif ((client_choice == 'tosh' and server_choice == 'qaychi') or
              (client_choice == 'qaychi' and server_choice == 'qog\'oz') or
              (client_choice == 'qog\'oz' and server_choice == 'tosh')):
            return f"Yutdingiz! Siz: {client_choice}, Server: {server_choice}"
        else:
            return f"Yutqazdingiz! Siz: {client_choice}, Server: {server_choice}"
    
    def log_request(self, message, addr):
        """Log requests to file"""
        try:
            with open(self.log_file, 'a', encoding='utf-8') as f:
                timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
                f.write(f"{timestamp} - {addr[0]}:{addr[1]} - {message}\n")
        except Exception as e:
            print(f"Log yozishda xatolik: {e}")
    
    def start_background_service(self):
        """Start background service that sends periodic messages"""
        def background_task():
            while self.running:
                time.sleep(10)  # Wait 10 seconds
                if self.connected_users and self.running:
                    message = f"Server xabari: {datetime.now().strftime('%H:%M:%S')}"
                    # Send to all connected users
                    for user_key, addr in list(self.connected_users.items()):
                        try:
                            self.socket.sendto(message.encode('utf-8'), addr)
                        except:
                            # Remove inactive users
                            if user_key in self.connected_users:
                                del self.connected_users[user_key]
        
        threading.Thread(target=background_task, daemon=True).start()
    
    def stop(self):
        """Stop the server"""
        self.running = False
        self.socket.close()

def main():
    """Main function to start the server"""
    server = UDPServer()
    try:
        server.start()
    except KeyboardInterrupt:
        print("\nServer to'xtatilmoqda...")
        server.stop()

if __name__ == "__main__":
    main()

File: udp/UDPServer.java
package udp;

import java.io.*;
import java.net.*;
import java.util.*;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * UDP Server with multiple functionalities
 * Implements various features from the project requirements
 */
class UDPServer {
    private static final int PORT = 9876;
    private static final int BUFFER_SIZE = 1024;
    private static final String LOG_FILE = "log.txt";
    
    // Store connected users
    private static Map<String, InetAddress> connectedUsers = new ConcurrentHashMap<>();
    private static Map<String, Integer> userPorts = new ConcurrentHashMap<>();
    
    // Database simulation
    private static Map<String, String> database = new HashMap<>();
    
    // Sensor data simulation
    private static Random random = new Random();
    
    // Scheduled executor for background services
    private static ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
    
    public static void main(String[] args) throws Exception {
        // Initialize database with sample data
        initializeDatabase();
        
        // Start background service
        startBackgroundService();
        
        DatagramSocket serverSocket = new DatagramSocket(PORT);
        byte[] receiveData = new byte[BUFFER_SIZE];
        byte[] sendData = new byte[BUFFER_SIZE];
        
        System.out.println("UDP Server ishga tushdi! Port: " + PORT);
        System.out.println("Server turli xil xizmatlarni qo'llab-quvvatlaydi:");
        System.out.println("- Oddiy xabar almashish");
        System.out.println("- Chat dasturi");
        System.out.println("- Matematik amallar");
        System.out.println("- Fayl mavjudligini tekshirish");
        System.out.println("- Tasodifiy son generatsiya");
        System.out.println("- Sensor ma'lumotlari");
        System.out.println("- Ma'lumotlar bazasi bilan ishlash");
        System.out.println("- Va boshqalar...");
        
        while (true) {
            try {
                DatagramPacket receivePacket = new DatagramPacket(receiveData, receiveData.length);
                serverSocket.receive(receivePacket);
                
                String clientMessage = new String(receivePacket.getData()).trim();
                InetAddress clientAddress = receivePacket.getAddress();
                int clientPort = receivePacket.getPort();
                
                // Log the request
                logRequest(clientMessage, clientAddress, clientPort);
                
                // Process the message and get response
                String response = processMessage(clientMessage, clientAddress, clientPort);
                
                // Send response back to client
                sendData = response.getBytes();
                DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, clientAddress, clientPort);
                serverSocket.send(sendPacket);
                
                System.out.println("Kliyent: " + clientAddress + ":" + clientPort + " - " + clientMessage);
                System.out.println("Javob: " + response);
                
            } catch (Exception e) {
                System.err.println("Xatolik: " + e.getMessage());
            }
        }
    }
    
    /**
     * Process incoming messages and return appropriate responses
     */
    private static String processMessage(String message, InetAddress clientAddress, int clientPort) {
        String response = "";
        
        // Register user
        String userKey = clientAddress.toString() + ":" + clientPort;
        connectedUsers.put(userKey, clientAddress);
        userPorts.put(userKey, clientPort);
        
        // Process different types of messages
        if (message.equalsIgnoreCase("salom") || message.equalsIgnoreCase("hello")) {
            response = "Salom! Xush kelibsiz! Server ishlamoqda.";
        }
        else if (message.equalsIgnoreCase("status")) {
            response = "Server ishlamoqda. Faol foydalanuvchilar: " + connectedUsers.size();
        }
        else if (message.equalsIgnoreCase("ip manzilim qanday?")) {
            response = "Sizning IP manzilingiz: " + clientAddress.getHostAddress();
        }
        else if (message.equalsIgnoreCase("foydalanuvchilar")) {
            response = "Faol foydalanuvchilar: " + connectedUsers.size() + " ta";
        }
        else if (message.equalsIgnoreCase("vaqt")) {
            response = "Hozirgi vaqt: " + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
        }
        else if (message.equalsIgnoreCase("tasodifiy son")) {
            int randomNum = random.nextInt(100) + 1;
            response = "Tasodifiy son (1-100): " + randomNum;
        }
        else if (message.equalsIgnoreCase("haroratni yuboring")) {
            int temperature = random.nextInt(40) + 10; // 10-50 degrees
            response = "Harorat: " + temperature + "°C";
        }
        else if (message.equalsIgnoreCase("menga maqol ayting")) {
            response = "Maqol: 'Bilim - kuchdir. Har bir kuni yangi narsa o'rganing.'";
        }
        else if (message.startsWith("fayl:")) {
            String fileName = message.substring(5).trim();
            File file = new File(fileName);
            if (file.exists()) {
                response = "Fayl '" + fileName + "' mavjud. Hajmi: " + file.length() + " bayt";
            } else {
                response = "Fayl '" + fileName + "' topilmadi";
            }
        }
        else if (message.startsWith("id:")) {
            String id = message.substring(3).trim();
            if (database.containsKey(id)) {
                response = "Ma'lumot (ID: " + id + "): " + database.get(id);
            } else {
                response = "ID '" + id + "' bo'yicha ma'lumot topilmadi";
            }
        }
        else if (message.equalsIgnoreCase("tosh") || message.equalsIgnoreCase("qaychi") || message.equalsIgnoreCase("qog'oz")) {
            String[] choices = {"tosh", "qaychi", "qog'oz"};
            String serverChoice = choices[random.nextInt(3)];
            response = playRockPaperScissors(message.toLowerCase(), serverChoice);
        }
        else if (message.contains("+") || message.contains("-") || message.contains("*") || message.contains("/")) {
            response = calculateMathExpression(message);
        }
        else if (message.equalsIgnoreCase("json")) {
            response = "{\"foydalanuvchi\":\"" + clientAddress.getHostAddress() + "\",\"vaqt\":\"" + 
                      LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")) + 
                      "\",\"status\":\"faol\"}";
        }
        else {
            // Default response - echo with timestamp
            response = "Sizning xabaringiz: '" + message + "' - " + 
                      LocalDateTime.now().format(DateTimeFormatter.ofPattern("HH:mm:ss"));
        }
        
        return response;
    }
    
    /**
     * Calculate mathematical expressions
     */
    private static String calculateMathExpression(String expression) {
        try {
            // Simple math expression parser
            if (expression.contains("+")) {
                String[] parts = expression.split("\\+");
                if (parts.length == 2) {
                    int a = Integer.parseInt(parts[0].trim());
                    int b = Integer.parseInt(parts[1].trim());
                    return "Natija: " + (a + b);
                }
            } else if (expression.contains("-")) {
                String[] parts = expression.split("-");
                if (parts.length == 2) {
                    int a = Integer.parseInt(parts[0].trim());
                    int b = Integer.parseInt(parts[1].trim());
                    return "Natija: " + (a - b);
                }
            } else if (expression.contains("*")) {
                String[] parts = expression.split("\\*");
                if (parts.length == 2) {
                    int a = Integer.parseInt(parts[0].trim());
                    int b = Integer.parseInt(parts[1].trim());
                    return "Natija: " + (a * b);
                }
            } else if (expression.contains("/")) {
                String[] parts = expression.split("/");
                if (parts.length == 2) {
                    int a = Integer.parseInt(parts[0].trim());
                    int b = Integer.parseInt(parts[1].trim());
                    if (b != 0) {
                        return "Natija: " + (a / b);
                    } else {
                        return "Xatolik: Nolga bo'lish mumkin emas";
                    }
                }
            }
        } catch (Exception e) {
            return "Xatolik: Noto'g'ri matematik ifoda";
        }
        return "Xatolik: Noto'g'ri format";
    }
    
    /**
     * Play Rock-Paper-Scissors game
     */
    private static String playRockPaperScissors(String clientChoice, String serverChoice) {
        if (clientChoice.equals(serverChoice)) {
            return "Durrang! Siz: " + clientChoice + ", Server: " + serverChoice;
        } else if ((clientChoice.equals("tosh") && serverChoice.equals("qaychi")) ||
                   (clientChoice.equals("qaychi") && serverChoice.equals("qog'oz")) ||
                   (clientChoice.equals("qog'oz") && serverChoice.equals("tosh"))) {
            return "Yutdingiz! Siz: " + clientChoice + ", Server: " + serverChoice;
        } else {
            return "Yutqazdingiz! Siz: " + clientChoice + ", Server: " + serverChoice;
        }
    }
    
    /**
     * Log requests to file
     */
    private static void logRequest(String message, InetAddress clientAddress, int clientPort) {
        try {
            FileWriter writer = new FileWriter(LOG_FILE, true);
            String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
            writer.write(timestamp + " - " + clientAddress + ":" + clientPort + " - " + message + "\n");
            writer.close();
        } catch (IOException e) {
            System.err.println("Log yozishda xatolik: " + e.getMessage());
        }
    }
    
    /**
     * Initialize sample database
     */
    private static void initializeDatabase() {
        database.put("1", "Ahmad Karimov - Dasturchi");
        database.put("2", "Malika Toshmatova - Dizayner");
        database.put("3", "Javlon Umarov - Menejer");
        database.put("4", "Dilnoza Rahimova - Marketing");
        database.put("5", "Bobur Ismoilov - Texnik");
    }
    
    /**
     * Start background service that sends periodic messages
     */
    private static void startBackgroundService() {
        scheduler.scheduleAtFixedRate(() -> {
            try {
                // Send periodic messages to all connected users
                if (!connectedUsers.isEmpty()) {
                    String message = "Server xabari: " + LocalDateTime.now().format(DateTimeFormatter.ofPattern("HH:mm:ss"));
                    byte[] data = message.getBytes();
                    
                    for (Map.Entry<String, InetAddress> entry : connectedUsers.entrySet()) {
                        try {
                            DatagramSocket socket = new DatagramSocket();
                            DatagramPacket packet = new DatagramPacket(
                                data, data.length, 
                                entry.getValue(), 
                                userPorts.get(entry.getKey())
                            );
                            socket.send(packet);
                            socket.close();
                        } catch (Exception e) {
                            // Remove inactive users
                            connectedUsers.remove(entry.getKey());
                            userPorts.remove(entry.getKey());
                        }
                    }
                }
            } catch (Exception e) {
                System.err.println("Background service xatoligi: " + e.getMessage());
            }
        }, 10, 10, TimeUnit.SECONDS);
    }
}
